U32 weight prediction

yancyribbens opened the pull request:
closes https://github.com/rust-bitcoin/rust-bitcoin/issues/4547

* Changes InputWeightPrediction to use `u32` types instead of `usize`
* Use saturating arithmetic to prevent overflows
* Remove 4Mwu bound from Arbitrary type

I tested these changes using libfuzzer and indeed InputWeightPrediction no longer overflows \o/

Kixunil commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1010,7 +1010,7 @@ impl InputWeightPrediction {
     ///
     /// [signature grinding]: https://bitcoin.stackexchange.com/questions/111660/what-is-signature-grinding
-    pub const fn ground_p2wpkh(bytes_to_grind: usize) -> Self {
+    pub const fn ground_p2wpkh(bytes_to_grind: u32) -> Self {
```
Notice that this is at most 62, thus it should've been `u8` from the start, not `usize`.

Kixunil commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1084,23 +1092,27 @@ impl InputWeightPrediction {
             let elem_len = witness_element_lengths[i];
-            let elem_size = elem_len + compact_size::encoded_size_const(elem_len as u64);
-            total_size += elem_size;
+            let encoded_size = compact_size::encoded_size_const(elem_len as u64) as u32;
```
Nope, you're just silently losing data here.

yancyribbens replied:
How is this silently losing data?  `encoded_size_const` can be at most 9, so we could cast as u4 even if there is such a thing.

```rust
#[inline]
pub const fn encoded_size_const(value: u64) -> usize {
    match value {
        0..=0xFC => 1,
        0xFD..=0xFFFF => 3,
        0x10000..=0xFFFFFFFF => 5,
        _ => 9,
    }
}
```

Kixunil replied:
Oh, sorry I misread it as element length directly. `encoded_size_const` MUST take `u64` but should return `u8`. It'll not help on it's own but perhaps we can have `const fn u8_to_usize` & co in `internals` to help with these const casts.

Kixunil commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1084,23 +1092,27 @@ impl InputWeightPrediction {
         let witness_size = if !witness_element_lengths.is_empty() {
-            total_size + compact_size::encoded_size_const(witness_element_lengths.len() as u64)
+            let encoded_size =
+                compact_size::encoded_size_const(witness_element_lengths.len() as u64) as u32;
```
And here.

Kixunil commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1084,23 +1092,27 @@ impl InputWeightPrediction {
-        let script_size =
-            input_script_len + compact_size::encoded_size_const(input_script_len as u64);
+
+        let encoded_size = compact_size::encoded_size_const(input_script_len as u64) as u32;
```
And here.

yancyribbens replied:
As above, this is at most 9 so I fail to see how this loses data..

Kixunil reviewed (changes requested):
Definitely not, this just silently corrupts data if someone were to feed it too large iterator or numbers. We could do it with panics which helps in finding "weird long slice" sooner but I'm not convinced we should use arbitrary limit of 4.2B rather than the natural limit of the platform.

yancyribbens commented:
> Definitely not, this just silently corrupts data if someone were to feed it too large iterator or numbers. 

I've addressed your comments.  I'm not sure if I am mistaken in someway, but as far as I can tell, `encoded_size_const` only returns at most 9 so I fail to see how casting this to `u32` in some way loses data.

> We could do it with panics which helps in finding "weird long slice" sooner but I'm not convinced we should use arbitrary limit of 4.2B rather than the natural limit of the platform.

I fail to see how a code workflow that includes panics is useful.  If anything, adding Result is better if you think these are limits that could in principle be hit during a workflow.  From my understanding, the only reason where it would be hit is during a "overshoot and prune" and if it panics well, there's nothing to prune.

yancyribbens commented:
> this just silently corrupts data if someone were to feed it too large iterator or numbers.

BTW, in order to hit the saturation point, one needs to pass an insanely large iterator that has more than u32::MAX number of elements.  It's so large you can't even write a test for it because the test frame times out when trying to build such a crazy large iterator.  I think this behavior should probably be documented though at least.

yancyribbensmobile commented:
>  Alternatively we could just saturate script_size (if I'm understanding correctly where the overflow is).

@tcharding sadly that won't work if you are suggesting saturating to u64::max.  The reason is that then if you call total_weight while using a script_size of u64::max, then you get another panic.

yancyribbens commented:
hmm on second thought, another possibility is to do saturating_add everywhere including in `total_weight` and then leave the types as `usize`.  That would probably go over better.

Kixunil commented:
You don't need insanely long iterator, it's enough that a single element is large (which can easily happen by miscomputation). Saturating everything sounds promising.

Kixunil reviewed (commented):
I'm still not convinced restricting the size is in any way useful but I appreciate new insights this PR brought.

Kixunil commented:
> How large of an element are we talking?

Just having one `usize::MAX`-large element should trigger the problem. But this made me realize, `InputWeightPrediction` needs to have `u64` internally to make the behavior consistent across platforms. It's converted to `u64` anyway, so doing it early makes sense.

yancyribbens commented:
> Just having one usize::MAX-large element should trigger the problem

That's no longer possible with this PR though

> But this made me realize, InputWeightPrediction needs to have u64 internally to make the behavior consistent across platforms. It's converted to u64 anyway, so doing it early makes sense.

Sounds like I should close this PR and make a new one that changes the internal types to `u64` and saturates to that.

yancyribbens commented:
I experimented with changing this to saturate at `u64`, however there is a lot of unsafe casts due to needing to go from usize to u64.  I'm not sure if there's a way we can clean this up more by using `u64` in more parts of the project instead of `usize`.

I started an issue here to discuss further: https://github.com/rust-bitcoin/rust-bitcoin/issues/4619

apoelstra commented:
These values are all sizes and we should use `usize`. Users will obtain them from length-calculating functions which all return `usize`.

In this crate (`primitives`, but not `units`) we assume that `usize` is always at least 32 bits and at most 64 bits. We have the `ToU64` trait from `internals` which codifies this, which we can use to convert usizes to u64s, though it will not work in a const context. (Personally I don't like this assumption and think we should avoid using it where we can. But sometimes we are more-or-less forced into it.)

In `InputWeightPrediction::new` we do a bunch of arithmetic which might overflow. As always, we have five choices:

* Panic (which we decided not to do, because the panic conditions are hard to describe and basically expose our exact logic)
* Wrap (lol definitely not)
* Return a `Result` (which I believe we want to avoid, since it just punts the panic to the user, and under sane conditions it will never happen).
* Use a bigger type or a tuple internally to do exact arithmetic (not worth the efficiency hit here IMO)
* Saturate

It seems like we've landed on the last option, which seems to me like the least bad of all of them. This is, after all, an input weight "prediction" and the user does not expect 100% precision, especially in cases where we give it ridiculously out-of-range values.

As you observe, just using the saturating arithmetic ops on `usize` is a bad user experience because it will lead to different results on 32-bit and 64-bit machines. (This is probably true all over the place when the user has `usize`s that are approaching 2^32, but ok, it would be nice if this particular function wasn't one of those places.)

With all this, **I propose** that we continue using `usize`s in the API, but saturate at `u32::MAX`. There are two ways to do this:  the simplest and most efficient way to do this is to use `u32`s internally and saturating ops, with private methods that convert to/from `usize`. The `usize`->`u32` map would saturate.

The `u32`->`usize` map is a bit more complicated -- we **could** just implement it as a cast with a comment saying that `usize` is always at least 32 bits. But I think it'd be more interesting to do `usize::try_from(x).unwrap_or(usize::MAX)` which will saturate on 16-bit machines and cast on normal ones. This has the nice effect that our logic works on machines of all word sizes, although our documentation (which will say that we saturate at 2^32) will be incorrect on 16-bit machines since the API will be exposing values that are saturated at 2^16.

Kixunil commented:
@apoelstra why do you think that having `u32` internally and saturating to that is better than having `u64`? We're losing precision here and the perf difference between 32 and 64 shouldn't be *that* large. But maybe it is and it's exactly what you're worried about?

apoelstra commented:
@Kixunil because `usize` may be smaller than `u64`, and we have methods that output `usize` (because we are outputting size estimates) and in those cases the observable behavior of the API will be different depending on the user's usize size.

yancyribbens commented:
> Hmm, actually maybe we should just do usize::try_from(x).unwrap_or(usize::MAX) like I suggested for u32, in order to technically support 16-bit machines in this module.

This turns into a larger change since function like [total_weight](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/transaction.rs#L1136) would need saturating addition also.  Although maybe that's more clear anyway to just do saturating add everywhere to `usize::MAX`.

yancyribbens commented:
Currently [total_weight](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/transaction.rs#L1136) won't overflow if usize is u64 but it would if `usize` was used internally instead of `u32`.

apoelstra commented:
Is there a difference between using `u64` internally and just using `usize` (and saturating `usize` arithmetic) internally?

Kixunil commented:
The difference is that `usize` allows different values on different architectures and `u32` does not. Though if you actually hit the number where it's different you'd have ~1GiB transaction even if you had zero non-segwit bytes and then you have way worse problem than that. Still, the accuracy is nice (but so is perf :man_shrugging: )

yancyribbens commented:
> Though if you actually hit the number where it's different you'd have ~1GiB transaction even if you had zero non-segwit bytes and then you have way worse problem than that

A 1GiB would never allowed in a block though?  `u16::MAX` seems like it would be too small to allow for an extreme script size, although I think `u32` MAX is theoretically large enough..

Kixunil commented:
For `u64`: greater precision - more information about how much it overflowed, for `u32`: better performance on MCUs. Anyway, now I think yes, `u32` is fine and if it becomes insufficient we can probably change it (with a flag).

yancyribbens commented:
Sounds good.  I think it might be worth changing some of the functions to saturate at `u32::MAX` so they don't overflow on 32 bit architectures.
